{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "197b563d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import json\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "from io import StringIO\n",
    "from Bio.PDB import PDBIO\n",
    "from Bio.PDB.NeighborSearch import NeighborSearch\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cfd7ea4",
   "metadata": {},
   "source": [
    "### Cell exploration - Cell 7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35425d54",
   "metadata": {},
   "source": [
    "### Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "edb8f1a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/29/w54hnxjs18q9cvq1pfzwym540000gn/T/ipykernel_22634/371685712.py:2: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gencode = pd.read_csv('data/GSE248049/annotation_files/Chlorocebus_sabaeus_genome.gff', sep='\\t', comment='#', header=None)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrName</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>key-value-pairs</th>\n",
       "      <th>other</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>126035930</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=NC_023642.1:1..126035930;Name=1;chromosome=...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>2628</td>\n",
       "      <td>4911</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-SCGB1C1;Dbxref=GeneID:103214198;Name=S...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>2628</td>\n",
       "      <td>4911</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-XM_007993457.1;Parent=gene-SCGB1C1;Dbxr...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>2628</td>\n",
       "      <td>3883</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-XM_007993457.1-1;Parent=rna-XM_0079934...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>4414</td>\n",
       "      <td>4911</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-XM_007993457.1-2;Parent=rna-XM_0079934...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrName  source feature  start        end score strand phase  \\\n",
       "0  NC_023642.1  RefSeq  region      1  126035930     .      +     .   \n",
       "1  NC_023642.1  Gnomon    gene   2628       4911     .      +     .   \n",
       "2  NC_023642.1  Gnomon    mRNA   2628       4911     .      +     .   \n",
       "3  NC_023642.1  Gnomon    exon   2628       3883     .      +     .   \n",
       "4  NC_023642.1  Gnomon    exon   4414       4911     .      +     .   \n",
       "\n",
       "                                     key-value-pairs other  \n",
       "0  ID=NC_023642.1:1..126035930;Name=1;chromosome=...     1  \n",
       "1  ID=gene-SCGB1C1;Dbxref=GeneID:103214198;Name=S...     1  \n",
       "2  ID=rna-XM_007993457.1;Parent=gene-SCGB1C1;Dbxr...     1  \n",
       "3  ID=exon-XM_007993457.1-1;Parent=rna-XM_0079934...     1  \n",
       "4  ID=exon-XM_007993457.1-2;Parent=rna-XM_0079934...     1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the GTF file using pandas\n",
    "gencode = pd.read_csv('data/GSE248049/annotation_files/Chlorocebus_sabaeus_genome.gff', sep='\\t', comment='#', header=None)\n",
    "\n",
    "# chromosome name, annonation source, feature type, genomic start loc, end loc, score, genomic strand, genomic phase\n",
    "# additional information as key-value pairs\n",
    "column_names = ['chrName', 'source', 'feature', 'start', 'end', 'score', 'strand', 'phase', 'key-value-pairs', \"other\"]\n",
    "gencode.columns = column_names\n",
    "\n",
    "# View the DataFrame\n",
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2ed42e44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=NC_023642.1:1..126035930;Name=1;chromosome=1;country=USA: NC%2C Wake Forest University;dev-stage=adult;gbkey=Src;genome=chromosome;isolate=1994-021;mol_type=genomic DNA;note=housed at the Wake Forest Primate Facility as part of the Vervet Research Colony;sex=male'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode.iloc[0]['key-value-pairs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3bfed8f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrName</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>key-value-pairs</th>\n",
       "      <th>other</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>2628</td>\n",
       "      <td>4911</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-SCGB1C1;Dbxref=GeneID:103214198;Name=S...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>4916</td>\n",
       "      <td>10379</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-ODF3;Dbxref=GeneID:103237012;Name=ODF3...</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>14423</td>\n",
       "      <td>18444</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-BET1L;Dbxref=GeneID:103216145;Name=BET...</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>18597</td>\n",
       "      <td>25398</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-RIC8A;Dbxref=GeneID:103222128;Name=RIC...</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>25740</td>\n",
       "      <td>49471</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-SIRT3;Dbxref=GeneID:103234991;Name=SIR...</td>\n",
       "      <td>1</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29346</th>\n",
       "      <td>NC_023672.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>4229367</td>\n",
       "      <td>4242824</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-LOC103246999;Dbxref=GeneID:103246999;N...</td>\n",
       "      <td>Y</td>\n",
       "      <td>1790656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29347</th>\n",
       "      <td>NC_023672.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>4486881</td>\n",
       "      <td>4489237</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-LOC103247001;Dbxref=GeneID:103247001;N...</td>\n",
       "      <td>Y</td>\n",
       "      <td>1790673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29348</th>\n",
       "      <td>NC_023672.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>4503583</td>\n",
       "      <td>4516685</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-LOC103247020;Dbxref=GeneID:103247020;N...</td>\n",
       "      <td>Y</td>\n",
       "      <td>1790679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29349</th>\n",
       "      <td>NC_023672.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>5352628</td>\n",
       "      <td>5364843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-LOC103247005;Dbxref=GeneID:103247005;N...</td>\n",
       "      <td>Y</td>\n",
       "      <td>1790696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29350</th>\n",
       "      <td>NC_023672.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>5636467</td>\n",
       "      <td>5638057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-LOC103247006;Dbxref=GeneID:103247006;N...</td>\n",
       "      <td>Y</td>\n",
       "      <td>1790714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29351 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chrName  source feature    start      end score strand phase  \\\n",
       "0      NC_023642.1  Gnomon    gene     2628     4911     .      +     .   \n",
       "1      NC_023642.1  Gnomon    gene     4916    10379     .      +     .   \n",
       "2      NC_023642.1  Gnomon    gene    14423    18444     .      -     .   \n",
       "3      NC_023642.1  Gnomon    gene    18597    25398     .      +     .   \n",
       "4      NC_023642.1  Gnomon    gene    25740    49471     .      -     .   \n",
       "...            ...     ...     ...      ...      ...   ...    ...   ...   \n",
       "29346  NC_023672.1  Gnomon    gene  4229367  4242824     .      -     .   \n",
       "29347  NC_023672.1  Gnomon    gene  4486881  4489237     .      -     .   \n",
       "29348  NC_023672.1  Gnomon    gene  4503583  4516685     .      -     .   \n",
       "29349  NC_023672.1  Gnomon    gene  5352628  5364843     .      +     .   \n",
       "29350  NC_023672.1  Gnomon    gene  5636467  5638057     .      +     .   \n",
       "\n",
       "                                         key-value-pairs other       id  \n",
       "0      ID=gene-SCGB1C1;Dbxref=GeneID:103214198;Name=S...     1        1  \n",
       "1      ID=gene-ODF3;Dbxref=GeneID:103237012;Name=ODF3...     1       10  \n",
       "2      ID=gene-BET1L;Dbxref=GeneID:103216145;Name=BET...     1       49  \n",
       "3      ID=gene-RIC8A;Dbxref=GeneID:103222128;Name=RIC...     1       60  \n",
       "4      ID=gene-SIRT3;Dbxref=GeneID:103234991;Name=SIR...     1      165  \n",
       "...                                                  ...   ...      ...  \n",
       "29346  ID=gene-LOC103246999;Dbxref=GeneID:103246999;N...     Y  1790656  \n",
       "29347  ID=gene-LOC103247001;Dbxref=GeneID:103247001;N...     Y  1790673  \n",
       "29348  ID=gene-LOC103247020;Dbxref=GeneID:103247020;N...     Y  1790679  \n",
       "29349  ID=gene-LOC103247005;Dbxref=GeneID:103247005;N...     Y  1790696  \n",
       "29350  ID=gene-LOC103247006;Dbxref=GeneID:103247006;N...     Y  1790714  \n",
       "\n",
       "[29351 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter rows where the feature attribute is 'gene'\n",
    "gene_rows = gencode[gencode['feature'] == 'gene'].copy()\n",
    "\n",
    "# Add the row number as 'id'\n",
    "gene_rows.loc[:, 'id'] = gene_rows.index\n",
    "\n",
    "# Reset the index to make it chronological and drop the old index\n",
    "gene_rows.reset_index(drop=True, inplace=True)\n",
    "\n",
    "gene_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3d15914f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=gene-SCGB1C1;Dbxref=GeneID:103214198;Name=SCGB1C1;gbkey=Gene;gene=SCGB1C1;gene_biotype=protein_coding'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_rows.iloc[0]['key-value-pairs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "15b6e175",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pdb data\n",
    "# from Bio.PDB.PDBParser import PDBParser\n",
    "folder_path = 'data/GSE80280/GSE80280-cell/'\n",
    "parser = PDBParser()\n",
    "genome_structure_model = parser.get_structure('genome_structure', folder_path+'cell7/GSM2219503_Cell_7_genome_structure_model.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba2425ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_id\n",
      "1530756    520\n",
      "1736805    400\n",
      "1856165    400\n",
      "1856670    400\n",
      "1683515    400\n",
      "          ... \n",
      "1758082     10\n",
      "1859332     10\n",
      "1596500     10\n",
      "1799374     10\n",
      "1596865      4\n",
      "Name: count, Length: 155, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# adding gene info to atoms and creating bonds\n",
    "node_link_dict = {}\n",
    "# Extract atom coordinates and other information\n",
    "for model in genome_structure_model:\n",
    "    atoms = []\n",
    "    nodes = []\n",
    "    links = []\n",
    "    \n",
    "    model_index = model.get_id()\n",
    "    \n",
    "    \n",
    "    # Create NeighborSearch object\n",
    "    ns = NeighborSearch(list(model.get_atoms()))\n",
    "    \n",
    "    \n",
    "    for chain in model:\n",
    "        for residue in chain:\n",
    "            for atom in residue:\n",
    "                atoms.append({\n",
    "                    'atom_name': atom.get_id(),\n",
    "                    'residue_name': residue.get_resname(),\n",
    "                    'residue': residue.get_id()[1] * 100000,\n",
    "                    'chain_id': atom.get_full_id()[2],\n",
    "                    'id': atom.get_serial_number(),\n",
    "                    'coords': atom.get_coord()\n",
    "                })\n",
    "\n",
    "    # Convert the atom information to a pandas DataFrame\n",
    "    df = pd.DataFrame(atoms)\n",
    "    \n",
    "    # add gene information\n",
    "    # Add a column for gene_id and initialize with None\n",
    "    df['gene_id'] = None\n",
    "    df['chrname'] = None\n",
    "    \n",
    "    # Iterate over gene_annotation to map atoms to genes\n",
    "    for index, row in gene_rows.iterrows():\n",
    "        gene_id = row['id']\n",
    "        chrname = row['chrName']\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "\n",
    "        # Map atoms to this gene\n",
    "        mask = (df['residue'] >= start) & (df['residue'] <= end)\n",
    "        df.loc[mask, 'gene_id'] = gene_id\n",
    "        df.loc[mask, 'chrname'] = chrname\n",
    "        \n",
    "# atom_name\tresidue_name\tresidue\tchain_id\tx\ty\tz\tid\tgene_id\tchrname\n",
    "    \n",
    "    print(df['gene_id'].value_counts())\n",
    "    for index, row in df.iterrows():\n",
    "        nodes.append({\n",
    "            \"id\": row['id'],\n",
    "            \"name\": row[\"atom_name\"],\n",
    "            \"residue_id\" : row[\"residue\"],\n",
    "            \"chain\": row[\"chain_id\"],\n",
    "            \"coord\": row['coords'].tolist(),\n",
    "            \"gene_id\": row[\"gene_id\"],\n",
    "            \"chrname\": row['chrname']\n",
    "        })\n",
    "        \n",
    "        # Find neighbors within a distance threshold - radius\n",
    "        neighbors = ns.search(row['coords'], 1.5)  # TODO : Adjust the radius as needed\n",
    "        \n",
    "        for neighbor in neighbors:\n",
    "            neighbor_id = neighbor.get_serial_number()\n",
    "            links.append({\"source\": row['id'], \"target\": neighbor_id})\n",
    "            \n",
    "    node_link_dict[model_index] = {\"nodes\": nodes, \"edges\": links}\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e57cd7f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "155"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "84857346",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "1530756    520\n",
       "1736805    400\n",
       "1856165    400\n",
       "1856670    400\n",
       "1683515    400\n",
       "          ... \n",
       "1758082     10\n",
       "1859332     10\n",
       "1596500     10\n",
       "1799374     10\n",
       "1596865      4\n",
       "Name: count, Length: 155, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c8a960f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node-link dictionary saved to genome_cell7_model0_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model1_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model2_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model3_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model4_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model5_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model6_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model7_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model8_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model9_gene_info.json\n"
     ]
    }
   ],
   "source": [
    "# for nd in node_link_dict.keys():\n",
    "#     json_file = f\"genome_cell7_model{nd}_gene_info.json\"\n",
    "#     filePath = 'data/extracted-node-link/cell7/'+json_file\n",
    "# #     print(json_file)\n",
    "#     with open(filePath, \"w\") as outfile:\n",
    "#         json.dump(node_link_dict[nd], outfile, indent=4)\n",
    "#     print(f\"Node-link dictionary saved to {json_file}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
