{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e50fc84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c302e134",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GFFProcessor:\n",
    "    def __init__(self, gff_file, seqId_to_chr_map_file = None):\n",
    "        self.gff_file = gff_file\n",
    "        self.seqId_to_chr_map_file = seqId_to_chr_map_file\n",
    "        self.df = None\n",
    "        self.chr_map_df = None\n",
    "        self.df_with_id_chromosome = None\n",
    "        self.gene_df = None\n",
    "        self.chromosome_df = None\n",
    "    \n",
    "    def load_gff_and_seq_to_chr_map_file(self):\n",
    "        # Read the GFF file into a pandas DataFrame\n",
    "        self.df = pd.read_csv(self.gff_file, sep='\\t', header=None, comment='#')\n",
    "        \n",
    "        # Determine the number of columns in the DataFrame\n",
    "        num_cols = len(self.df.columns)\n",
    "\n",
    "        # Define the column names based on the number of columns in the file\n",
    "        if num_cols == 9:\n",
    "            self.df.columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "        elif num_cols == 10:\n",
    "            self.df.columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes', 'add_col']\n",
    "        else:\n",
    "            raise ValueError(\"Column number < 9 or > 10 in GFF file\")\n",
    "            \n",
    "        if self.seqId_to_chr_map_file is not None:\n",
    "            self.chr_map_df = pd.read_csv(self.seqId_to_chr_map_file)\n",
    "#             print(self.chr_map_df.head())\n",
    "        \n",
    "        \n",
    "    def parse_attributes(self, attr_str):\n",
    "        # Function to parse GFF attributes into a dictionary\n",
    "        attributes = {}\n",
    "        items = attr_str.strip().split(';')\n",
    "        for item in items:\n",
    "            key_value = item.split('=')\n",
    "            if len(key_value) == 2:  # Ensure valid key-value pair\n",
    "                key = key_value[0].strip()\n",
    "                value = key_value[1].strip()\n",
    "                attributes[key] = value\n",
    "                        \n",
    "        return attributes\n",
    "    \n",
    "    def process_gff_data(self):\n",
    "        # Initialize lists to store gene data\n",
    "        genes = []\n",
    "        seqids = []\n",
    "        starts = []\n",
    "        ends = []\n",
    "        ids = []\n",
    "        chromosomes = []\n",
    "\n",
    "        # Iterate through each row in the DataFrame\n",
    "        for index, row in self.df.iterrows():\n",
    "            attributes = row['attributes']\n",
    "\n",
    "            # Parse attributes to get gene_id\n",
    "            attr_dict = self.parse_attributes(attributes)\n",
    "            if 'ID' in attr_dict:\n",
    "                if ':' in attr_dict['ID']:\n",
    "#                     print(attr_dict['ID'])\n",
    "                    gene_id = attr_dict['ID'].split(':')[0].strip()\n",
    "                    chromosome = attr_dict['chromosome'] if 'chromosome' in attr_dict else None\n",
    "                else:\n",
    "                    gene_id = attr_dict['ID']\n",
    "                    chromosome = None\n",
    "            else:\n",
    "                gene_id = ''\n",
    "                chromosome = None\n",
    "\n",
    "            # Append data to lists\n",
    "            genes.append(row['type'])\n",
    "            seqids.append(row['seqid'])\n",
    "            starts.append(row['start'])\n",
    "            ends.append(row['end'])\n",
    "            ids.append(gene_id)\n",
    "            chromosomes.append(chromosome)\n",
    "\n",
    "        # Create a new DataFrame with gene information\n",
    "        self.df_with_id_chromosome = pd.DataFrame({\n",
    "            'type': genes,\n",
    "            'seqid': seqids,\n",
    "            'start': starts,\n",
    "            'end': ends,\n",
    "            'id': ids,\n",
    "            'chromosome': chromosomes\n",
    "        })\n",
    "    \n",
    "#     def create_gene_df(self):\n",
    "#         # Filter DataFrame to include genes only\n",
    "#         self.gene_df = self.df_with_id_chromosome[self.df_with_id_chromosome['type'] == 'gene'][['seqid', 'start', 'end', 'id']]\n",
    "\n",
    "#         # Rename the index column to 'id'\n",
    "# #         self.gene_df.loc[:, 'id'] = self.gene_df.index\n",
    "\n",
    "#         # Create a column without \"gene-\"\n",
    "#         self.gene_df['name'] = self.gene_df['id'].str.replace('gene-', '')\n",
    "\n",
    "#         # Reset the index\n",
    "#         self.gene_df.reset_index(drop=True, inplace=True)\n",
    "        \n",
    "#     def create_chromosome_df(self):\n",
    "#         # filter df to include chromosomes only\n",
    "#         self.chromosome_df = self.df_with_id_chromosome[self.df_with_id_chromosome['chromosome'].notna()]\n",
    "        \n",
    "#         self.chromosome_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    def create_gene_df(self):\n",
    "        # Filter DataFrame to include genes only\n",
    "        self.gene_df = self.df_with_id_chromosome[self.df_with_id_chromosome['type'] == 'gene'][['seqid', 'start', 'end', 'id']]\n",
    "\n",
    "        # Create a column without \"gene-\"\n",
    "        self.gene_df['name'] = self.gene_df['id'].str.replace('gene-', '')\n",
    "\n",
    "        # If chr_map_df is provided, update seqid and store the old seqid\n",
    "        if self.chr_map_df is not None:\n",
    "            # Select only the necessary columns from chr_map_df\n",
    "            chr_map_subset = self.chr_map_df[['Contig', 'Chromosome']]\n",
    "\n",
    "            # Merge with chr_map_subset to get the Chromosome values\n",
    "            self.gene_df = self.gene_df.merge(\n",
    "                chr_map_subset, \n",
    "                how='left', \n",
    "                left_on='seqid', \n",
    "                right_on='Contig'\n",
    "            )\n",
    "\n",
    "            # Store the old seqid in seqid_old\n",
    "            self.gene_df['seqid_old'] = self.gene_df['seqid']\n",
    "\n",
    "            # Update seqid with Chromosome values\n",
    "            self.gene_df['seqid'] = self.gene_df['Chromosome']\n",
    "\n",
    "            # Drop the temporary columns from the merge\n",
    "            self.gene_df.drop(columns=['Contig', 'Chromosome'], inplace=True)\n",
    "\n",
    "        # Reset the index\n",
    "        self.gene_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    def create_chromosome_df(self):\n",
    "        # Filter df to include chromosomes only\n",
    "        self.chromosome_df = self.df_with_id_chromosome[self.df_with_id_chromosome['chromosome'].notna()]\n",
    "\n",
    "        # If chr_map_df is provided, update seqid and store the old seqid\n",
    "        if self.chr_map_df is not None:\n",
    "            chr_map_subset = self.chr_map_df[['Contig', 'Chromosome']]\n",
    "        \n",
    "            # Merge with chr_map_subset to get the Chromosome values\n",
    "            self.chromosome_df = self.chromosome_df.merge(\n",
    "                chr_map_subset, \n",
    "                how='left', \n",
    "                left_on='seqid', \n",
    "                right_on='Contig'\n",
    "            )\n",
    "\n",
    "            # Store the old seqid in seqid_old\n",
    "            self.chromosome_df['seqid_old'] = self.chromosome_df['seqid']\n",
    "\n",
    "            # Update seqid with Chromosome values\n",
    "            self.chromosome_df['seqid'] = self.chromosome_df['Chromosome']\n",
    "\n",
    "            # Drop the temporary columns from the merge\n",
    "            self.chromosome_df.drop(columns=['Contig', 'Chromosome'], inplace=True)\n",
    "\n",
    "        # Reset the index\n",
    "        self.chromosome_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    def display_gene_df(self):\n",
    "        # Display the gene DataFrame\n",
    "        if self.gene_df is not None:\n",
    "            print(self.gene_df.head())\n",
    "        else:\n",
    "            print(\"Gene DataFrame has not been created yet.\")\n",
    "            \n",
    "    def display_chromosome_df(self):\n",
    "        # Display the gene DataFrame\n",
    "        if self.chromosome_df is not None:\n",
    "            print(self.chromosome_df.head())\n",
    "        else:\n",
    "            print(\"Chromosome DataFrame has not been created yet.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2f43d7bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/29/w54hnxjs18q9cvq1pfzwym540000gn/T/ipykernel_22611/669737268.py:13: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  self.df = pd.read_csv(self.gff_file, sep='\\t', header=None, comment='#')\n"
     ]
    }
   ],
   "source": [
    "# Initialize the GFFProcessor instance with your GFF file path\n",
    "gff_processor = GFFProcessor('data/GSE248049/annotation_files/Chlorocebus_sabaeus_genome.gff', \n",
    "                            'data/GSE248049/annotation_files/Chlorocebus_sabeus_mva.genome.map.csv')\n",
    "\n",
    "# Load and process the GFF file\n",
    "gff_processor.load_gff_and_seq_to_chr_map_file()\n",
    "# extract important attributes\n",
    "gff_processor.process_gff_data()\n",
    "#create gene df\n",
    "gff_processor.create_gene_df()\n",
    "#create chromosome df\n",
    "gff_processor.create_chromosome_df()\n",
    "\n",
    "# gff_processor.display_chromosome_df()\n",
    "# gff_processor.display_gene_df()\n",
    "\n",
    "# Save the processed DataFrame to CSV\n",
    "gene_save_path = 'data/GSE248049/annotation_files/Chlorocebus_sabaeus_genome_genes.csv'\n",
    "gff_processor.gene_df.to_csv(gene_save_path, index=False)\n",
    "\n",
    "\n",
    "# Save the processed DataFrame to CSV\n",
    "chr_save_path = 'data/GSE248049/annotation_files/Chlorocebus_sabaeus_genome_chromosomes.csv'\n",
    "gff_processor.chromosome_df.to_csv(chr_save_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "09115734",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>region</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>1</td>\n",
       "      <td>126035930</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>2628</td>\n",
       "      <td>4911</td>\n",
       "      <td>gene-SCGB1C1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>2628</td>\n",
       "      <td>4911</td>\n",
       "      <td>rna-XM_007993457.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>2628</td>\n",
       "      <td>3883</td>\n",
       "      <td>exon-XM_007993457.1-1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>4414</td>\n",
       "      <td>4911</td>\n",
       "      <td>exon-XM_007993457.1-2</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type        seqid  start        end                     id chromosome\n",
       "0  region  NC_023642.1      1  126035930            NC_023642.1          1\n",
       "1    gene  NC_023642.1   2628       4911           gene-SCGB1C1       None\n",
       "2    mRNA  NC_023642.1   2628       4911     rna-XM_007993457.1       None\n",
       "3    exon  NC_023642.1   2628       3883  exon-XM_007993457.1-1       None\n",
       "4    exon  NC_023642.1   4414       4911  exon-XM_007993457.1-2       None"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_processor.df_with_id_chromosome.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e5280f82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>seqid_old</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2628</td>\n",
       "      <td>4911</td>\n",
       "      <td>gene-SCGB1C1</td>\n",
       "      <td>SCGB1C1</td>\n",
       "      <td>NC_023642.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4916</td>\n",
       "      <td>10379</td>\n",
       "      <td>gene-ODF3</td>\n",
       "      <td>ODF3</td>\n",
       "      <td>NC_023642.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14423</td>\n",
       "      <td>18444</td>\n",
       "      <td>gene-BET1L</td>\n",
       "      <td>BET1L</td>\n",
       "      <td>NC_023642.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18597</td>\n",
       "      <td>25398</td>\n",
       "      <td>gene-RIC8A</td>\n",
       "      <td>RIC8A</td>\n",
       "      <td>NC_023642.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25740</td>\n",
       "      <td>49471</td>\n",
       "      <td>gene-SIRT3</td>\n",
       "      <td>SIRT3</td>\n",
       "      <td>NC_023642.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid  start    end            id     name    seqid_old\n",
       "0  chr1   2628   4911  gene-SCGB1C1  SCGB1C1  NC_023642.1\n",
       "1  chr1   4916  10379     gene-ODF3     ODF3  NC_023642.1\n",
       "2  chr1  14423  18444    gene-BET1L    BET1L  NC_023642.1\n",
       "3  chr1  18597  25398    gene-RIC8A    RIC8A  NC_023642.1\n",
       "4  chr1  25740  49471    gene-SIRT3    SIRT3  NC_023642.1"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_processor.gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2e26d589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>seqid_old</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>region</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>126035930</td>\n",
       "      <td>NC_023642.1</td>\n",
       "      <td>1</td>\n",
       "      <td>NC_023642.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>region</td>\n",
       "      <td>chr10</td>\n",
       "      <td>1</td>\n",
       "      <td>128595539</td>\n",
       "      <td>NC_023651.1</td>\n",
       "      <td>10</td>\n",
       "      <td>NC_023651.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>region</td>\n",
       "      <td>chr11</td>\n",
       "      <td>1</td>\n",
       "      <td>128539186</td>\n",
       "      <td>NC_023652.1</td>\n",
       "      <td>11</td>\n",
       "      <td>NC_023652.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>region</td>\n",
       "      <td>chr12</td>\n",
       "      <td>1</td>\n",
       "      <td>108555830</td>\n",
       "      <td>NC_023653.1</td>\n",
       "      <td>12</td>\n",
       "      <td>NC_023653.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>region</td>\n",
       "      <td>chr13</td>\n",
       "      <td>1</td>\n",
       "      <td>98384682</td>\n",
       "      <td>NC_023654.1</td>\n",
       "      <td>13</td>\n",
       "      <td>NC_023654.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type  seqid  start        end           id chromosome    seqid_old\n",
       "0  region   chr1      1  126035930  NC_023642.1          1  NC_023642.1\n",
       "1  region  chr10      1  128595539  NC_023651.1         10  NC_023651.1\n",
       "2  region  chr11      1  128539186  NC_023652.1         11  NC_023652.1\n",
       "3  region  chr12      1  108555830  NC_023653.1         12  NC_023653.1\n",
       "4  region  chr13      1   98384682  NC_023654.1         13  NC_023654.1"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_processor.chromosome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "af3ffbfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculatePosition(pointA, pointB, percent):\n",
    "    # Calculate the differences in coordinates\n",
    "    delta_x = pointB[0] - pointA[0]\n",
    "    delta_y = pointB[1] - pointA[1]\n",
    "    delta_z = pointB[2] - pointA[2]\n",
    "\n",
    "    # Calculate % of the differences\n",
    "    dif_x = (percent / 100) * delta_x\n",
    "    dif_y = (percent / 100) * delta_y\n",
    "    dif_z = (percent / 100) * delta_z\n",
    "\n",
    "    # Calculate the coordinates of the point at % along the line\n",
    "    x_pos = pointA[0] + dif_x\n",
    "    y_pos = pointA[1] + dif_y\n",
    "    z_pos = pointA[2] + dif_z\n",
    "    \n",
    "    return [x_pos, y_pos, z_pos]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "824333e7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def gene_coord_extraction(df, df_gene, resolution):   \n",
    "    \n",
    "    #deep copy\n",
    "    gene_df = df_gene.copy(deep=True)\n",
    "    \n",
    "    #values to store\n",
    "    start_ids = []\n",
    "    end_ids = []\n",
    "    start_pos = []\n",
    "    start_x = []\n",
    "    start_y = []\n",
    "    start_z = []\n",
    "    end_pos = []\n",
    "    end_x = []\n",
    "    end_y = []\n",
    "    end_z = []\n",
    "    start_percent = []\n",
    "    end_percent = []\n",
    "    middle = []\n",
    "    middle_percent = []\n",
    "    middle_pos = []\n",
    "    middle_x = []\n",
    "    middle_y = []\n",
    "    middle_z = []\n",
    "    \n",
    "    #iterate over the gene data frame\n",
    "    for index, row in gene_df.iterrows():\n",
    "        #get the start and end position\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "        \n",
    "        #find the mid position\n",
    "        mid = int((start + end) / 2)\n",
    "        middle.append(mid)\n",
    "\n",
    "        # Calculate the ID range that encompasses the start and end\n",
    "        start_id = (start / resolution)\n",
    "        end_id = (end / resolution)\n",
    "\n",
    "        start_id_int = int(start_id)\n",
    "        end_id_int = int(end_id) + 1\n",
    "\n",
    "        start_ids.append(start_id_int)\n",
    "        end_ids.append(end_id_int)\n",
    "        \n",
    "        # find the percentage of where the mid point is located\n",
    "        m_percent = ((mid - (start_id_int * resolution)) / ((end_id_int * resolution) - (start_id_int * resolution))) * 100\n",
    "        s_percent = m_percent = ((start - (start_id_int * resolution)) / ((end_id_int * resolution) - (start_id_int * resolution))) * 100\n",
    "        e_percent = m_percent = ((end - (start_id_int * resolution)) / ((end_id_int * resolution) - (start_id_int * resolution))) * 100\n",
    "        \n",
    "        middle_percent.append(m_percent)\n",
    "        start_percent.append(s_percent)\n",
    "        end_percent.append(e_percent)\n",
    "\n",
    "#         print(start_id, end_id, start_id_int,end_id_int, len(df))\n",
    "\n",
    "        #getting the coordinates of the start and end beads that encompasses the gene\n",
    "        pointA = 0\n",
    "        pointB = 0\n",
    "        \n",
    "        # Check if start_id and end_id are within the bounds of df_structure\n",
    "        if (start_id_int <= len(df) - 1) and (end_id_int <= len(df) - 1):\n",
    "            \n",
    "#                         print(start_id_int, end_id_int, len(df))\n",
    "            start_row = df[df['id'] == start_id_int]\n",
    "            end_row = df[df['id'] == end_id_int]\n",
    "#             print(start_id_int, end_id_int, len(df))\n",
    "            # Extract x, y, z values if rows are found\n",
    "            if not start_row.empty:\n",
    "                pointA = [start_row.iloc[0]['x'], start_row.iloc[0]['y'], start_row.iloc[0]['z']]\n",
    "#                 start_pos.append(pointA)\n",
    "            if not end_row.empty:\n",
    "                pointB = [end_row.iloc[0]['x'], end_row.iloc[0]['y'], end_row.iloc[0]['z']]\n",
    "#                 end_pos.append(pointB)\n",
    "        else:\n",
    "            \n",
    "#             start_pos.append(None)\n",
    "#             end_pos.append(None)\n",
    "                                \n",
    "            pointA = None\n",
    "            pointB = None\n",
    "            \n",
    "#         print(pointA, pointB)\n",
    "            \n",
    "        #calcualate the coords of the mid point of the gene \n",
    "        # using parametric equation\n",
    "        # for two points P1, P2\n",
    "        # The parametric equation of the line passing through these two points can be written as:\n",
    "        # P(t) = P1 + t. (P2- P1) where t is the percentage value (0-1)\n",
    "        if pointA is not None and pointB is not None:\n",
    "#             print('not none')\n",
    "            sp = calculatePosition(pointA, pointB, s_percent)\n",
    "            mp = calculatePosition(pointA, pointB, m_percent)\n",
    "            ep = calculatePosition(pointA, pointB, e_percent)        \n",
    "            \n",
    "            start_pos.append(sp)\n",
    "            middle_pos.append(mp)\n",
    "            end_pos.append(ep)\n",
    "            start_x.append(sp[0])\n",
    "            start_y.append(sp[1])\n",
    "            start_z.append(sp[2])\n",
    "            end_x.append(ep[0])\n",
    "            end_y.append(ep[1])\n",
    "            end_z.append(ep[2])\n",
    "            middle_x.append(mp[0])\n",
    "            middle_y.append(mp[1])\n",
    "            middle_z.append(mp[2])\n",
    "            \n",
    "        else:\n",
    "            middle_pos.append(None)\n",
    "            start_pos.append(None)\n",
    "            end_pos.append(None)\n",
    "            start_x.append(None)\n",
    "            start_y.append(None)\n",
    "            start_z.append(None)\n",
    "            end_x.append(None)\n",
    "            end_y.append(None)\n",
    "            end_z.append(None)\n",
    "            middle_x.append(None)\n",
    "            middle_y.append(None)\n",
    "            middle_z.append(None)\n",
    "            \n",
    "\n",
    "\n",
    "    gene_df['start_id'] = start_ids\n",
    "    gene_df['end_id'] = end_ids\n",
    "    gene_df['start_pos'] = start_pos\n",
    "    gene_df['start_x'] = start_x\n",
    "    gene_df['start_y'] = start_y\n",
    "    gene_df['start_z'] = start_z\n",
    "    gene_df['end_pos'] = end_pos\n",
    "    gene_df['end_x'] = end_x\n",
    "    gene_df['end_y'] = end_y\n",
    "    gene_df['end_z'] = end_z\n",
    "    gene_df['middle'] = middle\n",
    "    gene_df['middle_x'] = middle_x\n",
    "    gene_df['middle_y'] = middle_y\n",
    "    gene_df['middle_z'] = middle_z\n",
    "    gene_df['middle_percent'] = middle_percent\n",
    "    gene_df['middle_pos'] = middle_pos\n",
    "                \n",
    "    print(\"calculation done\")       \n",
    "    return gene_df\n",
    "\n",
    "\n",
    "# gene_df_with_start_end_pos = gene_coord_extraction(bead_df, gff_processor.gene_df, 100000)\n",
    "\n",
    "# gene_df_with_start_end_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d5173351",
   "metadata": {},
   "outputs": [],
   "source": [
    "structure_file_path = 'data/GSE248049/all_structure_files'\n",
    "\n",
    "resolution = '100000'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "dc7ab5fd",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing folder: chr23\n",
      "chr_name 23\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr23/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr23/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr23/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr23/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr23/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr23/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr24\n",
      "chr_name 24\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr24/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr24/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr24/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr24/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr24/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr12\n",
      "chr_name 12\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr12/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr12/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr12/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr12/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr12/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr12/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr15\n",
      "chr_name 15\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr15/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr15/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr15/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr15/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr15/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr15/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr14\n",
      "chr_name 14\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr14/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr14/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr14/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr14/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr14/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr14/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr13\n",
      "chr_name 13\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr13/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr13/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr13/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr13/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr13/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr13/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr25\n",
      "chr_name 25\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr25/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr25/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr25/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr25/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr25/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr25/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr22\n",
      "chr_name 22\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr22/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr22/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr22/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr22/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr22/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr22/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr4\n",
      "chr_name 4\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr4/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr4/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr4/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr4/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr4/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr4/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr3\n",
      "chr_name 3\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr3/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr3/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr3/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr3/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr3/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr3/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr2\n",
      "chr_name 2\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr2/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr2/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr2/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr2/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr2/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr2/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr5\n",
      "chr_name 5\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr5/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr5/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr5/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr5/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr5/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr5/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr18\n",
      "chr_name 18\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr18/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr18/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr18/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr18/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr18/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data saved data/GSE248049/all_structure_files/chr18/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr27\n",
      "chr_name 27\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr27/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr27/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr27/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr27/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr27/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr27/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr20\n",
      "chr_name 20\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr20/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr20/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr20/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr20/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr20/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr20/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr29\n",
      "chr_name 29\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr29/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr29/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr29/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr29/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr29/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr29/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr16\n",
      "chr_name 16\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr16/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr16/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr16/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr16/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr16/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr16/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr11\n",
      "chr_name 11\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr11/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr11/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr11/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr11/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr11/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr10\n",
      "chr_name 10\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr10/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr10/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr10/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr10/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr10/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr10/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr28\n",
      "chr_name 28\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr28/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr28/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr28/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr28/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr28/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr28/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr17\n",
      "chr_name 17\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr17/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr17/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr17/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr17/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr17/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr21\n",
      "chr_name 21\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr21/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr21/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr21/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr21/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr21/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr21/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr19\n",
      "chr_name 19\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr19/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr19/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr19/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr19/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr19/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr19/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr26\n",
      "chr_name 26\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr26/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr26/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr26/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr26/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr26/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr26/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr7\n",
      "chr_name 7\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr7/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr7/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr7/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr7/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr7/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr7/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr9\n",
      "chr_name 9\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr9/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr9/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr9/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr9/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data saved data/GSE248049/all_structure_files/chr9/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr8\n",
      "chr_name 8\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr8/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr8/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr8/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr8/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr6\n",
      "chr_name 6\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr6/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr6/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr6/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr6/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr6/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr6/structure_18hrs_vacv_gene_info.csv\n",
      "Processing folder: chr1\n",
      "chr_name 1\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr1/structure_24hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr1/structure_18hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr1/structure_12hrs_untr_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr1/structure_24hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr1/structure_12hrs_vacv_gene_info.csv\n",
      "calculation done\n",
      "data saved data/GSE248049/all_structure_files/chr1/structure_18hrs_vacv_gene_info.csv\n"
     ]
    }
   ],
   "source": [
    "# Iterate through files in the section directory\n",
    "def bead_file_read_gene_coord_extract(atom_file_path, res):\n",
    "    \n",
    "    chr_folder_path = os.path.join(atom_file_path)\n",
    "    for chr_folder_name in os.listdir(chr_folder_path):\n",
    "        chr_path = os.path.join(chr_folder_path, chr_folder_name)\n",
    "#         print(chr_folder_name)\n",
    "\n",
    "        # Check if the current item is a directory starting with 'chr'\n",
    "        if os.path.isdir(chr_path) and chr_folder_name.startswith('chr'):  \n",
    "            print(f\"Processing folder: {chr_folder_name}\")\n",
    "            \n",
    "            # List all CSV files in the folder, considering only those already containing '_with_id0'\n",
    "            csv_files = [f for f in os.listdir(chr_path) if f.endswith('.csv') and '_with_id0' in f]\n",
    "            \n",
    "            chr_name = ''.join(filter(str.isdigit, chr_folder_name))  \n",
    "            \n",
    "            #removing the 0 at the first\n",
    "            chr_name = chr_name.lstrip('0')\n",
    "            \n",
    "            print(\"chr_name\", chr_name)\n",
    "\n",
    "            chr_gene_df = gff_processor.gene_df[gff_processor.gene_df['seqid'] == chr_folder_name]\n",
    "            \n",
    "            for csv_file in csv_files:\n",
    "                csv_file_path = os.path.join(chr_path, csv_file)\n",
    "\n",
    "                df = pd.read_csv(csv_file_path)\n",
    "                \n",
    "#                 print(res)\n",
    "                gene_df_with_start_end_pos = gene_coord_extraction(df, chr_gene_df, res)\n",
    "                \n",
    "                #rows_with_nan = gene_df_with_start_end_pos[pd.isna(gene_df_with_start_end_pos['middle_pos'])]\n",
    "                \n",
    "                \n",
    "                filtered_df = gene_df_with_start_end_pos.dropna(subset=['middle_pos'])\n",
    "                \n",
    "#                 print(\"gene df length\", len(gene_df_with_start_end_pos))\n",
    "#                 print(\"gene df length without NAN middle_pos\", len(filtered_df))\n",
    "\n",
    "                # Generate new output filename with \"_with_id0\"\n",
    "                base_name = csv_file.replace('_with_id0.csv', '')  # Remove \".csv\"\n",
    "                new_file_name = f\"{base_name}_gene_info.csv\"\n",
    "                save_file_path = os.path.join(chr_path, new_file_name)\n",
    "\n",
    "                filtered_df.to_csv(save_file_path, index=False)\n",
    "                \n",
    "                print(f\"data saved {save_file_path}\")\n",
    "\n",
    "            \n",
    "    \n",
    "\n",
    "            \n",
    "bead_file_read_gene_coord_extract(structure_file_path, 100000)\n",
    "# bead_file_read_gene_coord_extract(after_path, time_hr, resolution, structure_file_name, 250000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4202bdb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba9a0a4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c823f2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
