{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e50fc84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c302e134",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GFFProcessor:\n",
    "    def __init__(self, gff_file):\n",
    "        self.gff_file = gff_file\n",
    "        self.df = None\n",
    "        self.df_with_id_chromosome = None\n",
    "        self.gene_df = None\n",
    "        self.chromosome_df = None\n",
    "    \n",
    "    def load_gff_file(self):\n",
    "        # Read the GFF file into a pandas DataFrame\n",
    "        self.df = pd.read_csv(self.gff_file, sep='\\t', header=None, comment='#')\n",
    "        \n",
    "        # Determine the number of columns in the DataFrame\n",
    "        num_cols = len(self.df.columns)\n",
    "\n",
    "        # Define the column names based on the number of columns in the file\n",
    "        if num_cols == 9:\n",
    "            self.df.columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "        elif num_cols == 10:\n",
    "            self.df.columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes', 'add_col']\n",
    "        else:\n",
    "            raise ValueError(\"Column number < 9 or > 10 in GFF file\")\n",
    "        \n",
    "        \n",
    "    def parse_attributes(self, attr_str):\n",
    "        # Function to parse GFF attributes into a dictionary\n",
    "        attributes = {}\n",
    "        items = attr_str.strip().split(';')\n",
    "        for item in items:\n",
    "            key_value = item.split('=')\n",
    "            if len(key_value) == 2:  # Ensure valid key-value pair\n",
    "                key = key_value[0].strip()\n",
    "                value = key_value[1].strip()\n",
    "                attributes[key] = value\n",
    "                        \n",
    "        return attributes\n",
    "    \n",
    "    def process_gff_data(self):\n",
    "        # Initialize lists to store gene data\n",
    "        genes = []\n",
    "        seqids = []\n",
    "        starts = []\n",
    "        ends = []\n",
    "        ids = []\n",
    "        chromosomes = []\n",
    "\n",
    "        # Iterate through each row in the DataFrame\n",
    "        for index, row in self.df.iterrows():\n",
    "            attributes = row['attributes']\n",
    "\n",
    "            # Parse attributes to get gene_id\n",
    "            attr_dict = self.parse_attributes(attributes)\n",
    "            if 'ID' in attr_dict:\n",
    "                if ':' in attr_dict['ID']:\n",
    "#                     print(attr_dict['ID'])\n",
    "                    gene_id = attr_dict['ID'].split(':')[0].strip()\n",
    "                    chromosome = attr_dict['chromosome'] if 'chromosome' in attr_dict else None\n",
    "                else:\n",
    "                    gene_id = attr_dict['ID']\n",
    "                    chromosome = None\n",
    "            else:\n",
    "                gene_id = ''\n",
    "                chromosome = None\n",
    "\n",
    "            # Append data to lists\n",
    "            genes.append(row['type'])\n",
    "            seqids.append(row['seqid'])\n",
    "            starts.append(row['start'])\n",
    "            ends.append(row['end'])\n",
    "            ids.append(gene_id)\n",
    "            chromosomes.append(chromosome)\n",
    "\n",
    "        # Create a new DataFrame with gene information\n",
    "        self.df_with_id_chromosome = pd.DataFrame({\n",
    "            'type': genes,\n",
    "            'seqid': seqids,\n",
    "            'start': starts,\n",
    "            'end': ends,\n",
    "            'id': ids,\n",
    "            'chromosome': chromosomes\n",
    "        })\n",
    "    \n",
    "    def create_gene_df(self):\n",
    "        # Filter DataFrame to include genes only\n",
    "        self.gene_df = self.df_with_id_chromosome[self.df_with_id_chromosome['type'] == 'gene'][['seqid', 'start', 'end', 'id']]\n",
    "\n",
    "        # Rename the index column to 'id'\n",
    "#         self.gene_df.loc[:, 'id'] = self.gene_df.index\n",
    "\n",
    "        # Create a column without \"gene-\"\n",
    "        self.gene_df['name'] = self.gene_df['id'].str.replace('gene-', '')\n",
    "\n",
    "        # Reset the index\n",
    "        self.gene_df.reset_index(drop=True, inplace=True)\n",
    "        \n",
    "    def create_chromosome_df(self):\n",
    "        # filter df to include chromosomes only\n",
    "        self.chromosome_df = self.df_with_id_chromosome[self.df_with_id_chromosome['chromosome'].notna()]\n",
    "        \n",
    "        self.chromosome_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    def display_gene_df(self):\n",
    "        # Display the gene DataFrame\n",
    "        if self.gene_df is not None:\n",
    "            print(self.gene_df.head())\n",
    "        else:\n",
    "            print(\"Gene DataFrame has not been created yet.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2f43d7bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize the GFFProcessor instance with your GFF file path\n",
    "gff_processor = GFFProcessor('data/MRC5/gencode_raw/GCF_009914755.1_T2T-CHM13v2.0_genomic.named.gff')\n",
    "\n",
    "# Load and process the GFF file\n",
    "gff_processor.load_gff_file()\n",
    "# extract important attributes\n",
    "gff_processor.process_gff_data()\n",
    "#create gene df\n",
    "gff_processor.create_gene_df()\n",
    "#create chromosome df\n",
    "gff_processor.create_chromosome_df()\n",
    "\n",
    "# Save the processed DataFrame to CSV\n",
    "gene_save_path = 'data/MRC5/gencode_raw/gene-chromosome/GCF_009914755.1_T2T-CHM13v2.0_genomic.named_genes.csv'\n",
    "gff_processor.gene_df.to_csv(gene_save_path, index=False)\n",
    "\n",
    "\n",
    "# Save the processed DataFrame to CSV\n",
    "chr_save_path = 'data/MRC5/gencode_raw/gene-chromosome/GCF_009914755.1_T2T-CHM13v2.0_genomic.named_chromosomes.csv'\n",
    "gff_processor.chromosome_df.to_csv(chr_save_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "09115734",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>region</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>248387328</td>\n",
       "      <td>NC_060925.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7506</td>\n",
       "      <td>138480</td>\n",
       "      <td>gene-LOC127239154</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>7506</td>\n",
       "      <td>138480</td>\n",
       "      <td>rna-NR_182074.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>138321</td>\n",
       "      <td>138480</td>\n",
       "      <td>exon-NR_182074.1-1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>129906</td>\n",
       "      <td>130010</td>\n",
       "      <td>exon-NR_182074.1-2</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type seqid   start        end                  id chromosome\n",
       "0   region  chr1       1  248387328         NC_060925.1          1\n",
       "1     gene  chr1    7506     138480   gene-LOC127239154       None\n",
       "2  lnc_RNA  chr1    7506     138480     rna-NR_182074.1       None\n",
       "3     exon  chr1  138321     138480  exon-NR_182074.1-1       None\n",
       "4     exon  chr1  129906     130010  exon-NR_182074.1-2       None"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_processor.df_with_id_chromosome.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e5280f82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7506</td>\n",
       "      <td>138480</td>\n",
       "      <td>gene-LOC127239154</td>\n",
       "      <td>LOC127239154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20892</td>\n",
       "      <td>23710</td>\n",
       "      <td>gene-LOC124905685</td>\n",
       "      <td>LOC124905685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>52979</td>\n",
       "      <td>54612</td>\n",
       "      <td>gene-LOC101928626</td>\n",
       "      <td>LOC101928626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>111940</td>\n",
       "      <td>112877</td>\n",
       "      <td>gene-OR4F16</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>152269</td>\n",
       "      <td>205171</td>\n",
       "      <td>gene-LOC100288069</td>\n",
       "      <td>LOC100288069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41504</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62159112</td>\n",
       "      <td>62204700</td>\n",
       "      <td>gene-SPRY3-2</td>\n",
       "      <td>SPRY3-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41505</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62306325</td>\n",
       "      <td>62368760</td>\n",
       "      <td>gene-VAMP7-2</td>\n",
       "      <td>VAMP7-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41506</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62422543</td>\n",
       "      <td>62435805</td>\n",
       "      <td>gene-IL9R-2</td>\n",
       "      <td>IL9R-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41507</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62439553</td>\n",
       "      <td>62441822</td>\n",
       "      <td>gene-WASIR1-2</td>\n",
       "      <td>WASIR1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41508</th>\n",
       "      <td>chrY</td>\n",
       "      <td>62446549</td>\n",
       "      <td>62449392</td>\n",
       "      <td>gene-LOC124908905-2</td>\n",
       "      <td>LOC124908905-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41509 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid     start       end                   id            name\n",
       "0      chr1      7506    138480    gene-LOC127239154    LOC127239154\n",
       "1      chr1     20892     23710    gene-LOC124905685    LOC124905685\n",
       "2      chr1     52979     54612    gene-LOC101928626    LOC101928626\n",
       "3      chr1    111940    112877          gene-OR4F16          OR4F16\n",
       "4      chr1    152269    205171    gene-LOC100288069    LOC100288069\n",
       "...     ...       ...       ...                  ...             ...\n",
       "41504  chrY  62159112  62204700         gene-SPRY3-2         SPRY3-2\n",
       "41505  chrY  62306325  62368760         gene-VAMP7-2         VAMP7-2\n",
       "41506  chrY  62422543  62435805          gene-IL9R-2          IL9R-2\n",
       "41507  chrY  62439553  62441822        gene-WASIR1-2        WASIR1-2\n",
       "41508  chrY  62446549  62449392  gene-LOC124908905-2  LOC124908905-2\n",
       "\n",
       "[41509 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_processor.gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e26d589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>region</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>248387328</td>\n",
       "      <td>NC_060925.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>region</td>\n",
       "      <td>chr2</td>\n",
       "      <td>1</td>\n",
       "      <td>242696752</td>\n",
       "      <td>NC_060926.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>region</td>\n",
       "      <td>chr3</td>\n",
       "      <td>1</td>\n",
       "      <td>201105948</td>\n",
       "      <td>NC_060927.1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>region</td>\n",
       "      <td>chr4</td>\n",
       "      <td>1</td>\n",
       "      <td>193574945</td>\n",
       "      <td>NC_060928.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>region</td>\n",
       "      <td>chr5</td>\n",
       "      <td>1</td>\n",
       "      <td>182045439</td>\n",
       "      <td>NC_060929.1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     type seqid  start        end           id chromosome\n",
       "0  region  chr1      1  248387328  NC_060925.1          1\n",
       "1  region  chr2      1  242696752  NC_060926.1          2\n",
       "2  region  chr3      1  201105948  NC_060927.1          3\n",
       "3  region  chr4      1  193574945  NC_060928.1          4\n",
       "4  region  chr5      1  182045439  NC_060929.1          5"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_processor.chromosome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d5173351",
   "metadata": {},
   "outputs": [],
   "source": [
    "before_path = 'data/MRC5/processed/MRC5/mock'\n",
    "after_path = 'data/MRC5/processed/MRC5/229E'\n",
    "time_hr = '48hr'\n",
    "resolution = '250000'\n",
    "structure_file_name = 'structure-with-id0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "af3ffbfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculatePosition(pointA, pointB, percent):\n",
    "    # Calculate the differences in coordinates\n",
    "    delta_x = pointB[0] - pointA[0]\n",
    "    delta_y = pointB[1] - pointA[1]\n",
    "    delta_z = pointB[2] - pointA[2]\n",
    "\n",
    "    # Calculate % of the differences\n",
    "    dif_x = (percent / 100) * delta_x\n",
    "    dif_y = (percent / 100) * delta_y\n",
    "    dif_z = (percent / 100) * delta_z\n",
    "\n",
    "    # Calculate the coordinates of the point at % along the line\n",
    "    x_pos = pointA[0] + dif_x\n",
    "    y_pos = pointA[1] + dif_y\n",
    "    z_pos = pointA[2] + dif_z\n",
    "    \n",
    "    return [x_pos, y_pos, z_pos]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "824333e7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def gene_coord_extraction(df, df_gene, resolution):   \n",
    "    \n",
    "    #deep copy\n",
    "    gene_df = df_gene.copy(deep=True)\n",
    "    \n",
    "    #values to store\n",
    "    start_ids = []\n",
    "    end_ids = []\n",
    "    start_pos = []\n",
    "    start_x = []\n",
    "    start_y = []\n",
    "    start_z = []\n",
    "    end_pos = []\n",
    "    end_x = []\n",
    "    end_y = []\n",
    "    end_z = []\n",
    "    start_percent = []\n",
    "    end_percent = []\n",
    "    middle = []\n",
    "    middle_percent = []\n",
    "    middle_pos = []\n",
    "    middle_x = []\n",
    "    middle_y = []\n",
    "    middle_z = []\n",
    "    \n",
    "    #iterate over the gene data frame\n",
    "    for index, row in gene_df.iterrows():\n",
    "        #get the start and end position\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "        \n",
    "        #find the mid position\n",
    "        mid = int((start + end) / 2)\n",
    "        middle.append(mid)\n",
    "\n",
    "        # Calculate the ID range that encompasses the start and end\n",
    "        start_id = (start / resolution)\n",
    "        end_id = (end / resolution)\n",
    "\n",
    "        start_id_int = int(start_id)\n",
    "        end_id_int = int(end_id) + 1\n",
    "\n",
    "        start_ids.append(start_id_int)\n",
    "        end_ids.append(end_id_int)\n",
    "        \n",
    "        # find the percentage of where the mid point is located\n",
    "        m_percent = ((mid - (start_id_int * resolution)) / ((end_id_int * resolution) - (start_id_int * resolution))) * 100\n",
    "        s_percent = m_percent = ((start - (start_id_int * resolution)) / ((end_id_int * resolution) - (start_id_int * resolution))) * 100\n",
    "        e_percent = m_percent = ((end - (start_id_int * resolution)) / ((end_id_int * resolution) - (start_id_int * resolution))) * 100\n",
    "        \n",
    "        middle_percent.append(m_percent)\n",
    "        start_percent.append(s_percent)\n",
    "        end_percent.append(e_percent)\n",
    "\n",
    "#         print(start_id, end_id, start_id_int,end_id_int, len(df))\n",
    "\n",
    "        #getting the coordinates of the start and end beads that encompasses the gene\n",
    "        pointA = 0\n",
    "        pointB = 0\n",
    "        \n",
    "        # Check if start_id and end_id are within the bounds of df_structure\n",
    "        if (start_id_int <= len(df) - 1) and (end_id_int <= len(df) - 1):\n",
    "            \n",
    "#                         print(start_id_int, end_id_int, len(df))\n",
    "            start_row = df[df['id'] == start_id_int]\n",
    "            end_row = df[df['id'] == end_id_int]\n",
    "#             print(start_id_int, end_id_int, len(df))\n",
    "            # Extract x, y, z values if rows are found\n",
    "            if not start_row.empty:\n",
    "                pointA = [start_row.iloc[0]['x'], start_row.iloc[0]['y'], start_row.iloc[0]['z']]\n",
    "#                 start_pos.append(pointA)\n",
    "            if not end_row.empty:\n",
    "                pointB = [end_row.iloc[0]['x'], end_row.iloc[0]['y'], end_row.iloc[0]['z']]\n",
    "#                 end_pos.append(pointB)\n",
    "        else:\n",
    "            \n",
    "#             start_pos.append(None)\n",
    "#             end_pos.append(None)\n",
    "                                \n",
    "            pointA = None\n",
    "            pointB = None\n",
    "            \n",
    "#         print(pointA, pointB)\n",
    "            \n",
    "        #calcualate the coords of the mid point of the gene \n",
    "        # using parametric equation\n",
    "        # for two points P1, P2\n",
    "        # The parametric equation of the line passing through these two points can be written as:\n",
    "        # P(t) = P1 + t. (P2- P1) where t is the percentage value (0-1)\n",
    "        if pointA is not None and pointB is not None:\n",
    "#             print('not none')\n",
    "            sp = calculatePosition(pointA, pointB, s_percent)\n",
    "            mp = calculatePosition(pointA, pointB, m_percent)\n",
    "            ep = calculatePosition(pointA, pointB, e_percent)        \n",
    "            \n",
    "            start_pos.append(sp)\n",
    "            middle_pos.append(mp)\n",
    "            end_pos.append(ep)\n",
    "            start_x.append(sp[0])\n",
    "            start_y.append(sp[1])\n",
    "            start_z.append(sp[2])\n",
    "            end_x.append(ep[0])\n",
    "            end_y.append(ep[1])\n",
    "            end_z.append(ep[2])\n",
    "            middle_x.append(mp[0])\n",
    "            middle_y.append(mp[1])\n",
    "            middle_z.append(mp[2])\n",
    "            \n",
    "        else:\n",
    "            middle_pos.append(None)\n",
    "            start_pos.append(None)\n",
    "            end_pos.append(None)\n",
    "            start_x.append(None)\n",
    "            start_y.append(None)\n",
    "            start_z.append(None)\n",
    "            end_x.append(None)\n",
    "            end_y.append(None)\n",
    "            end_z.append(None)\n",
    "            middle_x.append(None)\n",
    "            middle_y.append(None)\n",
    "            middle_z.append(None)\n",
    "            \n",
    "\n",
    "\n",
    "    gene_df['start_id'] = start_ids\n",
    "    gene_df['end_id'] = end_ids\n",
    "    gene_df['start_pos'] = start_pos\n",
    "    gene_df['start_x'] = start_x\n",
    "    gene_df['start_y'] = start_y\n",
    "    gene_df['start_z'] = start_z\n",
    "    gene_df['end_pos'] = end_pos\n",
    "    gene_df['end_x'] = end_x\n",
    "    gene_df['end_y'] = end_y\n",
    "    gene_df['end_z'] = end_z\n",
    "    gene_df['middle'] = middle\n",
    "    gene_df['middle_x'] = middle_x\n",
    "    gene_df['middle_y'] = middle_y\n",
    "    gene_df['middle_z'] = middle_z\n",
    "    gene_df['middle_percent'] = middle_percent\n",
    "    gene_df['middle_pos'] = middle_pos\n",
    "                \n",
    "    print(\"calculation done\")       \n",
    "    return gene_df\n",
    "\n",
    "\n",
    "# gene_df_with_start_end_pos = gene_coord_extraction(bead_df, gff_processor.gene_df, 100000)\n",
    "\n",
    "# gene_df_with_start_end_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dc7ab5fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".DS_Store\n",
      "chr10\n",
      "chr_name 10\n",
      "chr_id ['NC_060934.1']\n",
      "calculation done\n",
      "gene df length 1592\n",
      "gene df length without NAN middle_pos 1592\n",
      "data saved data/MRC5/processed/MRC5/mock/48hr/250000/chr10/gene-info.csv\n",
      "chr10 okay\n",
      "chr17\n",
      "chr_name 17\n",
      "chr_id ['NC_060941.1']\n",
      "calculation done\n",
      "gene df length 1944\n",
      "gene df length without NAN middle_pos 1943\n",
      "data saved data/MRC5/processed/MRC5/mock/48hr/250000/chr17/gene-info.csv\n",
      "chr17 okay\n",
      "chr21\n",
      "chr_name 21\n",
      "chr_id ['NC_060945.1']\n",
      "calculation done\n",
      "gene df length 775\n",
      "gene df length without NAN middle_pos 775\n",
      "data saved data/MRC5/processed/MRC5/mock/48hr/250000/chr21/gene-info.csv\n",
      "chr21 okay\n",
      "chr19\n",
      "chr_name 19\n",
      "chr_id ['NC_060943.1']\n",
      "calculation done\n",
      "gene df length 2029\n",
      "gene df length without NAN middle_pos 2029\n",
      "data saved data/MRC5/processed/MRC5/mock/48hr/250000/chr19/gene-info.csv\n",
      "chr19 okay\n",
      ".DS_Store\n",
      "chr10\n",
      "chr_name 10\n",
      "chr_id ['NC_060934.1']\n",
      "calculation done\n",
      "gene df length 1592\n",
      "gene df length without NAN middle_pos 1592\n",
      "data saved data/MRC5/processed/MRC5/229E/48hr/250000/chr10/gene-info.csv\n",
      "chr10 okay\n",
      "chr17\n",
      "chr_name 17\n",
      "chr_id ['NC_060941.1']\n",
      "calculation done\n",
      "gene df length 1944\n",
      "gene df length without NAN middle_pos 1944\n",
      "data saved data/MRC5/processed/MRC5/229E/48hr/250000/chr17/gene-info.csv\n",
      "chr17 okay\n",
      "chr21\n",
      "chr_name 21\n",
      "chr_id ['NC_060945.1']\n",
      "calculation done\n",
      "gene df length 775\n",
      "gene df length without NAN middle_pos 775\n",
      "data saved data/MRC5/processed/MRC5/229E/48hr/250000/chr21/gene-info.csv\n",
      "chr21 okay\n",
      "chr19\n",
      "chr_name 19\n",
      "chr_id ['NC_060943.1']\n",
      "calculation done\n",
      "gene df length 2029\n",
      "gene df length without NAN middle_pos 2029\n",
      "data saved data/MRC5/processed/MRC5/229E/48hr/250000/chr19/gene-info.csv\n",
      "chr19 okay\n"
     ]
    }
   ],
   "source": [
    "# Iterate through files in the section directory\n",
    "def bead_file_read_gene_coord_extract(atom_file_path, hour, resolution, structure_file_name, res):\n",
    "    chr_folder_path = os.path.join(atom_file_path, hour, resolution)\n",
    "    for chr_folder_name in os.listdir(chr_folder_path):\n",
    "        print(chr_folder_name)\n",
    "\n",
    "        # Check if the current item is a directory starting with 'chr'\n",
    "        if os.path.isdir(chr_folder_path) and chr_folder_name.startswith('chr'):        \n",
    "            # Navigate into the chr folder and look for structure-with-tracks.csv\n",
    "#             csv_file_path = os.path.join(chr_folder_path, 'structure', '100kb', 'structure-with-id0.csv')\n",
    "            \n",
    "            csv_file_path = os.path.join(chr_folder_path, chr_folder_name, f\"{structure_file_name}.csv\" )\n",
    "            \n",
    "            chr_name = ''.join(filter(str.isdigit, chr_folder_name))  \n",
    "            \n",
    "            #removing the 0 at the first\n",
    "            chr_name = chr_name.lstrip('0')\n",
    "            \n",
    "            print(\"chr_name\", chr_name)\n",
    "            \n",
    "            chr_id = gff_processor.chromosome_df.loc[gff_processor.chromosome_df['chromosome'] == chr_name, 'id'].tolist()\n",
    "            print(\"chr_id\", chr_id)\n",
    "            \n",
    "            # human seqID is the chromosomes\n",
    "            # TODO dynamically fix this\n",
    "#             chr_gene_df = gff_processor.gene_df[gff_processor.gene_df['seqid'] == chr_id[0]]\n",
    "\n",
    "            chr_gene_df = gff_processor.gene_df[gff_processor.gene_df['seqid'] == chr_folder_name]\n",
    "            \n",
    "#             print(chr_gene_df.head())\n",
    "\n",
    "            if os.path.exists(csv_file_path):\n",
    "                df = pd.read_csv(csv_file_path)\n",
    "                \n",
    "#                 print(res)\n",
    "                gene_df_with_start_end_pos = gene_coord_extraction(df, chr_gene_df, res)\n",
    "                \n",
    "                #rows_with_nan = gene_df_with_start_end_pos[pd.isna(gene_df_with_start_end_pos['middle_pos'])]\n",
    "                \n",
    "                \n",
    "                filtered_df = gene_df_with_start_end_pos.dropna(subset=['middle_pos'])\n",
    "                \n",
    "                print(\"gene df length\", len(gene_df_with_start_end_pos))\n",
    "                print(\"gene df length without NAN middle_pos\", len(filtered_df))\n",
    "            \n",
    "#                 csv_file_path = f\"data/MRC5/gencode/{struc_type}/{chr_folder_name}-{struc_type}-gene-info.csv\"\n",
    "                csv_file_path = os.path.join(chr_folder_path, chr_folder_name, 'gene-info.csv')\n",
    "                filtered_df.to_csv(csv_file_path, index=False)\n",
    "                \n",
    "                print(f\"data saved {csv_file_path}\")\n",
    "                print(f\"{chr_folder_name} okay\")\n",
    "            \n",
    "    \n",
    "\n",
    "            \n",
    "bead_file_read_gene_coord_extract(before_path, time_hr, resolution, structure_file_name, 250000)\n",
    "bead_file_read_gene_coord_extract(after_path, time_hr, resolution, structure_file_name, 250000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4202bdb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba9a0a4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c823f2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
