{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "197b563d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import json\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "from io import StringIO\n",
    "from Bio.PDB import PDBIO\n",
    "from Bio.PDB.NeighborSearch import NeighborSearch\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cfd7ea4",
   "metadata": {},
   "source": [
    "### Cell exploration - Cell 7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35425d54",
   "metadata": {},
   "source": [
    "### Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "edb8f1a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrName</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>key-value-pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrName   source     feature    start      end score strand phase  \\\n",
       "0    chr1   HAVANA        gene  3143476  3144545     .      +     .   \n",
       "1    chr1   HAVANA  transcript  3143476  3144545     .      +     .   \n",
       "2    chr1   HAVANA        exon  3143476  3144545     .      +     .   \n",
       "3    chr1  ENSEMBL        gene  3172239  3172348     .      +     .   \n",
       "4    chr1  ENSEMBL  transcript  3172239  3172348     .      +     .   \n",
       "\n",
       "                                     key-value-pairs  \n",
       "0  gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...  \n",
       "1  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "2  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "3  gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...  \n",
       "4  gene_id \"ENSMUSG00000064842.3\"; transcript_id ...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the GTF file using pandas\n",
    "gencode = pd.read_csv('data/GSE80280/gencode.vM29.annotation.gtf', sep='\\t', comment='#', header=None)\n",
    "\n",
    "# chromosome name, annonation source, feature type, genomic start loc, end loc, score, genomic strand, genomic phase\n",
    "# additional information as key-value pairs\n",
    "column_names = ['chrName', 'source', 'feature', 'start', 'end', 'score', 'strand', 'phase', 'key-value-pairs']\n",
    "gencode.columns = column_names\n",
    "\n",
    "# View the DataFrame\n",
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3bfed8f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrName</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>key-value-pairs</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3276124</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.6\"; gene_type \"pro...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3322980</td>\n",
       "      <td>3323459</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102851.2\"; gene_type \"pro...</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3435954</td>\n",
       "      <td>3438772</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000103377.2\"; gene_type \"TEC...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55409</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>837364</td>\n",
       "      <td>840451</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095523.2\"; gene_type \"pro...</td>\n",
       "      <td>1869994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55410</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>910289</td>\n",
       "      <td>913083</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095475.2\"; gene_type \"pro...</td>\n",
       "      <td>1870005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55411</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>921942</td>\n",
       "      <td>924675</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094855.2\"; gene_type \"pro...</td>\n",
       "      <td>1870016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55412</th>\n",
       "      <td>JH584303.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>81607</td>\n",
       "      <td>82689</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095019.2\"; gene_type \"pro...</td>\n",
       "      <td>1870027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55413</th>\n",
       "      <td>JH584304.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>52190</td>\n",
       "      <td>59690</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095041.8\"; gene_type \"pro...</td>\n",
       "      <td>1870032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55414 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrName   source feature    start      end score strand phase  \\\n",
       "0            chr1   HAVANA    gene  3143476  3144545     .      +     .   \n",
       "1            chr1  ENSEMBL    gene  3172239  3172348     .      +     .   \n",
       "2            chr1   HAVANA    gene  3276124  3741721     .      -     .   \n",
       "3            chr1   HAVANA    gene  3322980  3323459     .      +     .   \n",
       "4            chr1   HAVANA    gene  3435954  3438772     .      -     .   \n",
       "...           ...      ...     ...      ...      ...   ...    ...   ...   \n",
       "55409  JH584299.1  ENSEMBL    gene   837364   840451     .      +     .   \n",
       "55410  JH584299.1  ENSEMBL    gene   910289   913083     .      -     .   \n",
       "55411  JH584299.1  ENSEMBL    gene   921942   924675     .      +     .   \n",
       "55412  JH584303.1  ENSEMBL    gene    81607    82689     .      +     .   \n",
       "55413  JH584304.1  ENSEMBL    gene    52190    59690     .      -     .   \n",
       "\n",
       "                                         key-value-pairs       id  \n",
       "0      gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...        0  \n",
       "1      gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...        3  \n",
       "2      gene_id \"ENSMUSG00000051951.6\"; gene_type \"pro...        6  \n",
       "3      gene_id \"ENSMUSG00000102851.2\"; gene_type \"pro...       24  \n",
       "4      gene_id \"ENSMUSG00000103377.2\"; gene_type \"TEC...       27  \n",
       "...                                                  ...      ...  \n",
       "55409  gene_id \"ENSMUSG00000095523.2\"; gene_type \"pro...  1869994  \n",
       "55410  gene_id \"ENSMUSG00000095475.2\"; gene_type \"pro...  1870005  \n",
       "55411  gene_id \"ENSMUSG00000094855.2\"; gene_type \"pro...  1870016  \n",
       "55412  gene_id \"ENSMUSG00000095019.2\"; gene_type \"pro...  1870027  \n",
       "55413  gene_id \"ENSMUSG00000095041.8\"; gene_type \"pro...  1870032  \n",
       "\n",
       "[55414 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter rows where the feature attribute is 'gene'\n",
    "gene_rows = gencode[gencode['feature'] == 'gene'].copy()\n",
    "\n",
    "# Add the row number as 'id'\n",
    "gene_rows.loc[:, 'id'] = gene_rows.index\n",
    "\n",
    "# Reset the index to make it chronological and drop the old index\n",
    "gene_rows.reset_index(drop=True, inplace=True)\n",
    "\n",
    "gene_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "15b6e175",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pdb data\n",
    "# from Bio.PDB.PDBParser import PDBParser\n",
    "folder_path = 'data/GSE80280/GSE80280-cell/'\n",
    "parser = PDBParser()\n",
    "genome_structure_model = parser.get_structure('genome_structure', folder_path+'cell7/GSM2219503_Cell_7_genome_structure_model.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba2425ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_id\n",
      "1530756    520\n",
      "1736805    400\n",
      "1856165    400\n",
      "1856670    400\n",
      "1683515    400\n",
      "          ... \n",
      "1758082     10\n",
      "1859332     10\n",
      "1596500     10\n",
      "1799374     10\n",
      "1596865      4\n",
      "Name: count, Length: 155, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# adding gene info to atoms and creating bonds\n",
    "node_link_dict = {}\n",
    "# Extract atom coordinates and other information\n",
    "for model in genome_structure_model:\n",
    "    atoms = []\n",
    "    nodes = []\n",
    "    links = []\n",
    "    \n",
    "    model_index = model.get_id()\n",
    "    \n",
    "    \n",
    "    # Create NeighborSearch object\n",
    "    ns = NeighborSearch(list(model.get_atoms()))\n",
    "    \n",
    "    \n",
    "    for chain in model:\n",
    "        for residue in chain:\n",
    "            for atom in residue:\n",
    "                atoms.append({\n",
    "                    'atom_name': atom.get_id(),\n",
    "                    'residue_name': residue.get_resname(),\n",
    "                    'residue': residue.get_id()[1] * 100000,\n",
    "                    'chain_id': atom.get_full_id()[2],\n",
    "                    'id': atom.get_serial_number(),\n",
    "                    'coords': atom.get_coord()\n",
    "                })\n",
    "\n",
    "    # Convert the atom information to a pandas DataFrame\n",
    "    df = pd.DataFrame(atoms)\n",
    "    \n",
    "    # add gene information\n",
    "    # Add a column for gene_id and initialize with None\n",
    "    df['gene_id'] = None\n",
    "    df['chrname'] = None\n",
    "    \n",
    "    # Iterate over gene_annotation to map atoms to genes\n",
    "    for index, row in gene_rows.iterrows():\n",
    "        gene_id = row['id']\n",
    "        chrname = row['chrName']\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "\n",
    "        # Map atoms to this gene\n",
    "        mask = (df['residue'] >= start) & (df['residue'] <= end)\n",
    "        df.loc[mask, 'gene_id'] = gene_id\n",
    "        df.loc[mask, 'chrname'] = chrname\n",
    "        \n",
    "# atom_name\tresidue_name\tresidue\tchain_id\tx\ty\tz\tid\tgene_id\tchrname\n",
    "    \n",
    "    print(df['gene_id'].value_counts())\n",
    "    for index, row in df.iterrows():\n",
    "        nodes.append({\n",
    "            \"id\": row['id'],\n",
    "            \"name\": row[\"atom_name\"],\n",
    "            \"residue_id\" : row[\"residue\"],\n",
    "            \"chain\": row[\"chain_id\"],\n",
    "            \"coord\": row['coords'].tolist(),\n",
    "            \"gene_id\": row[\"gene_id\"],\n",
    "            \"chrname\": row['chrname']\n",
    "        })\n",
    "        \n",
    "        # Find neighbors within a distance threshold - radius\n",
    "        neighbors = ns.search(row['coords'], 1.5)  # TODO : Adjust the radius as needed\n",
    "        \n",
    "        for neighbor in neighbors:\n",
    "            neighbor_id = neighbor.get_serial_number()\n",
    "            links.append({\"source\": row['id'], \"target\": neighbor_id})\n",
    "            \n",
    "    node_link_dict[model_index] = {\"nodes\": nodes, \"edges\": links}\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e57cd7f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "155"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "84857346",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "1530756    520\n",
       "1736805    400\n",
       "1856165    400\n",
       "1856670    400\n",
       "1683515    400\n",
       "          ... \n",
       "1758082     10\n",
       "1859332     10\n",
       "1596500     10\n",
       "1799374     10\n",
       "1596865      4\n",
       "Name: count, Length: 155, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c8a960f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node-link dictionary saved to genome_cell7_model0_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model1_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model2_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model3_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model4_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model5_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model6_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model7_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model8_gene_info.json\n",
      "Node-link dictionary saved to genome_cell7_model9_gene_info.json\n"
     ]
    }
   ],
   "source": [
    "# for nd in node_link_dict.keys():\n",
    "#     json_file = f\"genome_cell7_model{nd}_gene_info.json\"\n",
    "#     filePath = 'data/extracted-node-link/cell7/'+json_file\n",
    "# #     print(json_file)\n",
    "#     with open(filePath, \"w\") as outfile:\n",
    "#         json.dump(node_link_dict[nd], outfile, indent=4)\n",
    "#     print(f\"Node-link dictionary saved to {json_file}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
